A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91*99.

Find the largest palindrome made from the product of two -digit numbers.

## My solution

In [35]:
def check_palyndrome(number):
    number_str = str(number)
    number_str_list = list(number_str)
    start = 0
    end = len(number_str_list) - 1
    is_palyndrome = False
    while (number_str_list[start] == number_str_list[end] and start < end ):
        start +=1
        end -=1
    if start > end:
        is_palyndrome = True

    return is_palyndrome

check_palyndrome(9009), check_palyndrome(9019)

(True, False)

In [64]:
# num_1 = 999
palyndrome_list = []
for num_1 in range (999,99,-1):
    for num_2 in range (999,99,-1):
        number = num_1*num_2
        if check_palyndrome(number):
            print(num_1, num_2, number, check_palyndrome(number))
            palyndrome_list.append(number)

995 583 580085 True
995 517 514415 True
993 913 906609 True
991 121 119911 True
987 286 282282 True
987 143 141141 True
982 869 853358 True
979 664 650056 True
979 614 601106 True
979 605 592295 True
979 555 543345 True
979 496 485584 True
979 446 436634 True
979 387 378873 True
979 337 329923 True
978 418 408804 True
978 209 204402 True
977 638 623326 True
975 539 525525 True
974 308 299992 True
971 484 469964 True
968 916 886688 True
968 831 804408 True
968 722 698896 True
968 652 631136 True
968 637 616616 True
968 458 443344 True
968 443 428824 True
968 279 270072 True
968 264 255552 True
966 858 828828 True
966 429 414414 True
964 506 487784 True
963 187 180081 True
962 924 888888 True
962 693 666666 True
962 462 444444 True
962 231 222222 True
961 363 348843 True
959 803 770077 True
958 682 653356 True
957 894 855558 True
957 844 807708 True
957 728 696696 True
957 678 648846 True
957 555 531135 True
957 512 489984 True
957 389 372273 True
957 339 324423 True
957 173 165561 True


396 752 297792 True
396 707 279972 True
396 682 270072 True
396 637 252252 True
396 592 234432 True
396 547 216612 True
393 704 276672 True
393 407 159951 True
392 561 219912 True
391 682 266662 True
391 341 133331 True
389 957 372273 True
388 759 294492 True
387 979 378873 True
386 737 284482 True
381 341 129921 True
377 616 232232 True
376 792 297792 True
374 728 272272 True
374 713 266662 True
374 618 231132 True
374 603 225522 True
374 588 219912 True
372 726 270072 True
371 572 212212 True
369 539 198891 True
369 319 117711 True
368 649 238832 True
364 803 292292 True
364 748 272272 True
364 693 252252 True
364 638 232232 True
364 583 212212 True
363 961 348843 True
363 744 270072 True
363 704 255552 True
363 554 201102 True
363 447 162261 True
363 407 147741 True
357 616 219912 True
357 473 168861 True
352 846 297792 True
352 786 276672 True
352 726 255552 True
352 666 234432 True
352 606 213312 True
347 836 290092 True
341 883 301103 True
341 792 270072 True
341 782 266662 True


In [67]:
max(palyndrome_list)

906609

In [68]:
max_index = palyndrome_list.index(max(palyndrome_list))
max_index

2

## Official solution n°1

Let our palindrome be P = ab with a and b the two 3-digit numbers. 
If a and b are 3-digits long then they must lie between 100 and 999 inclusive. So an initial solution to the problem might be:

In [69]:
def reverse(n):
    reversed_num = 0
    while n > 0:
        reversed_num = 10*reversed_num + n%10
        n = n//10
    return reversed_num

reverse(1912)

2191

In [70]:
def isPalindrome(n):
    return n == reverse(n)

isPalindrome(9009), isPalindrome(9019), 

(True, False)

In [72]:
largestPalindrome = 0
a = 100
while a <= 999:
    b = 100
    while b <= 999:
        if isPalindrome(a*b) and a*b > largestPalindrome:
            largestPalindrome = a*b
        b = b+1
    a = a+1
largestPalindrome

906609

## Official solution n°2

Solution n°1 is fast enough for this case but could be improved. 

For starters, the current method checks many numbers multiple times. 
For example thenumber 69696 is checked both when a=132 and b=528 and when a=528 and b=132. 
To stop checking numbers like this we can assume a ≤ b, roughly halving the number of calculations needed.

In [73]:
largestPalindrome = 0
a = 100
while a <= 999:
    b = a
    while b <= 999:
        if isPalindrome(a*b) and a*b > largestPalindrome:
            largestPalindrome = a*b
        b = b+1
    a = a+1
largestPalindrome

906609

## Official solution n°3

Next we should consider counting downwards from 999 instead of counting upwards from 100. This makes the program far more likely to find a large palindrome earlier, and we can quite easily stop checking a and b that would be too small to improve upon the largest palindrome found so far.


In [75]:
largestPalindrome = 0
a = 999
while a >= 100:
    b = 999
    while b >= a:
        if  a*b <= largestPalindrome:
            break
        if isPalindrome(a*b):
            largestPalindrome = a*b
        b = b-1
    a = a-1
largestPalindrome

906609

This is fast but can be sped up further with some analysis. Consider the digits of P.
let them be x, y and z. P must be at least 6 digits long since the palindrome 111111 = 143×777 – the product of two 3-digit integers. 
Since P is palindromic:
- P=100000x + 10000y + 1000z + 100z + 10y + x
- P=100001x + 10010y + 1100z
- P=11*(9091x + 910y + 100z)

Since 11 is prime, at least one of the integers a or b must have a factor of 11.
So if a is not divisible by 11 then we know b must be. Using this information we can determine what values of b we check depending on a:

In [77]:
largestPalindrome = 0
a = 999
while a >= 100:
    if a % 11 == 0:
        b = 999
        db = 1
    else:
        b = 990 # //The largest number less than or equal 999 //and divisible by 11
        db = 11
    while b >= a:
        if a*b <= largestPalindrome:
            break
        if isPalindrome(a*b):
            largestPalindrome = a*b
        b = b-db
    a = a-1
largestPalindrome

906609